### DISCUSSION TITLE
Refreshing credentials used by already instantiated clients

### DISCUSSION THREAD

**Discussion by archevel** [aws/aws-sdk-net-extensions-cognito#124] [Q&A] [answered]
I am using credentials from Cognito with a few different AWS services (eg. CloudWatch, SQS and SecretsManager). The processen are long running and as such eventually the tokens expire and I need to use the refresh token to get new credentials. However, when this happens I will already have instantiated the clients with the cognito credentials. Using the cloudwatch client as an example:  

```c#
private static void ConfigureSerilogLogger(LoggerConfiguration configuration, CognitoAWSCredentials credentials, RegionEndpoint region)
{
    var client = new AmazonCloudWatchLogsClient(credentials, region);
    configuration
        // ... some additional log config
        .WriteTo.AmazonCloudWatch(
            cloudWatchClient: client,
            logGroup: "SomeLogGroup",
            logStreamPrefix: "SomeStream", 
            createLogGroup: false,
            appendUniqueInstanceGuid: false,
            appendHostName: false,
            textFormatter: new ExpressionTemplate(logOutputTemplate))
}
``` 

Here the `client` is passed to Serilog and it is a bit cumbersome to reinitialize the logger with new credentials.  Ideally the code using the cloudwatch client shouldn't need to care about the state of the credentials, but as is I would need to introduce a mechanism for recreating each client when I use the refresh token to get new credentials. Is there a way around this? I noticed that `CognitoAWSCredentials` inherit from `RefreshingAWSCredentials` and as a developer I would expect them to have some support for using a refresh token to generate new credentials before the expiry of id/access tokens. Maybe the SDK already have a mechanism for handling this gracefully?

**Comment by ashishdhingra** [comment, marked as answer]
@archevel The tokens received via Cognito have the `ExpirationTime`. The logic to refresh the tokens and reinitialize existing clients should be build within the application, may be using timer to be triggered based on value of expiration time. The StackOverflow post https://stackoverflow.com/questions/54938342/how-to-update-token-using-refresh-token-in-mvc-client-application throws some idea for ASP.NET MVC application, there might be better solutions available though. This library would not refresh token automatically.

[a comment by an automated account is not shown]
